lnc/lnc.go

```diff
@@ -0,0 +1,169 @@
+package lnc
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/lightninglabs/lightning-node-connect/mailbox"
+	"github.com/lightningnetwork/lnd/keychain"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/mwitkow/grpc-proxy/proxy"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// HeaderMacaroon is the HTTP header field name that is used to send
+// the macaroon.
+const HeaderMacaroon = "Macaroon"
+
+type conn struct {
+	client     lnrpc.LightningClient
+	grpcClient *grpc.ClientConn
+	creds      credentials.PerRPCCredentials
+	cancel     func()
+}
+
+func (c *conn) Close() error {
+	err := c.grpcClient.Close()
+	c.cancel()
+	return err
+}
+
+type LNC struct {
+	store Store
+
+	passphrase []byte
+	conn       *conn
+	mu         sync.Mutex
+}
+
+func NewLNC(store Store) *LNC {
+	return &LNC{
+		store: store,
+		mu:    sync.Mutex{},
+	}
+}
+
+func (l *LNC) OpenConn(ctx context.Context, passphrase []byte) error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if l.conn != nil {
+		return fmt.Errorf("connection already open")
+	}
+
+	session, err := l.store.GetSessionByPassphrase(ctx, passphrase)
+	if err != nil {
+		return err
+	}
+
+	conn, err := l.newConn(session)
+	if err != nil {
+		return err
+	}
+
+	l.conn = conn
+	l.passphrase = passphrase
+
+	return nil
+}
+
+func (l *LNC) CloseConn() error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
```

Comment:
One thing we'll want to test is that if the connection dies then we either retry, or we shut down.

Reply:
If the connection dies, the invoice subscription returns an error, and we shutdown the server (in the same way we do if you are directly connecting to your LND node)